subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultMutableAttributeContainer.java

```diff
@@ -16,102 +16,115 @@
 
 package org.gradle.api.internal.attributes;
 
-import com.google.common.collect.ImmutableSet;
-import com.google.common.collect.Sets;
 import org.gradle.api.attributes.Attribute;
 import org.gradle.api.attributes.AttributeContainer;
+import org.gradle.api.internal.provider.MappingProvider;
+import org.gradle.api.internal.provider.PropertyFactory;
 import org.gradle.api.internal.provider.ProviderInternal;
+import org.gradle.api.provider.MapProperty;
 import org.gradle.api.provider.Provider;
 import org.gradle.internal.Cast;
 import org.gradle.internal.isolation.Isolatable;
 
 import javax.annotation.Nullable;
-import java.util.Collections;
 import java.util.Comparator;
-import java.util.LinkedHashMap;
-import java.util.LinkedHashSet;
+import java.util.HashMap;
 import java.util.Map;
 import java.util.Objects;
 import java.util.Set;
 import java.util.TreeMap;
 
 final class DefaultMutableAttributeContainer extends AbstractAttributeContainer implements AttributeContainerInternal {
-    private final Map<Attribute<?>, Isolatable<?>> attributes = new LinkedHashMap<>(); // Need to maintain insertion order here, this is indirectly tested
-    private Map<Attribute<?>, Provider<?>> lazyAttributes = Cast.uncheckedCast(Collections.EMPTY_MAP);
-    private boolean realizingAttributes = false;
 
     private final DefaultAttributesFactory attributesFactory;
 
-    private ImmutableAttributes immutableValue;
+    private final MapProperty<Attribute<?>, Isolatable<?>> state;
+    private boolean realizingAttributes;
 
-    public DefaultMutableAttributeContainer(DefaultAttributesFactory attributesFactory) {
+    public DefaultMutableAttributeContainer(DefaultAttributesFactory attributesFactory, PropertyFactory propertyFactory) {
         this.attributesFactory = attributesFactory;
+        this.state = Cast.uncheckedNonnullCast(propertyFactory.mapProperty(Attribute.class, Isolatable.class));
     }
 
     @Override
     public String toString() {
         final Map<Attribute<?>, Object> sorted = new TreeMap<>(Comparator.comparing(Attribute::getName));
-        lazyAttributes.keySet().forEach(key -> sorted.put(key, lazyAttributes.get(key).toString()));
-        attributes.keySet().forEach(key -> sorted.put(key, attributes.get(key).toString()));
+        sorted.putAll(getRealizedMap());
```

Comment:
💅 Sigh, it's ironic to insist on insertion order then throw it away when `toString()`-ing.

Reply:
Where do we _insist_ on insertion order?

I'm not sure how significant the insertion order is here -- nothing on the interface of `AttributeContainer` claims we maintain insertion order, and the only collections exposed by the `AttributeContainer` interface is a _Set_, which explicitly does not enforce insertion order. 

In the dependency resolution engine, insertion order is not important. 

In practice, insertion order is sort of required during mutation, as older things must be overridden by newer mutable actions (if I say attribute foo = bar, then I say foo = baz, yeah I expect my later call to override my previous)

But nowhere do we say `keySet` will return in that order

Reply:
If you are referring to the test `lazy attributes are evaluated in insertion order`, that is about _evaluation_ of lazy code, but does not enforce at all how we return those values to the user